canonical/kfp-operators
# tests: enable bundle functional tests for v2 compiled pipelines

The recent upgrades in kubeflow pipelines introduced a new version of the SDK that is compatible with the newer backend (pipelines 2.0.3). This commit ensures that the CI tests both a pipeline v1 and v2 to ensure good coverage.

According to the [Version Compatibility](https://www.kubeflow.org/docs/components/pipelines/v2/version-compatibility/) table, pipelines that are compiled using the 1.8 SDK are still compatible with the 2.0 backend, hence the need to cover both cases in the CI for our solution. 
From the official documentation (and from experimentation) we can also tell that to be able to create runs from a compiled v1 and v2 pipelines using the Python SDK, their corresponding SDK versions must be used; that is, creating a v2 pipeline run can only be done with the v2 SDK, and the same applies for v1 compiled pipelines.

> NOTE: this PR proposes a change in the deletion method for an existing pipeline. The new backend expects the pipeline version to be deleted before the actual pipeline can be deleted; otherwise the backend will complain with the following msg:

```
HTTP response headers: HTTPHeaderDict({'X-Powered-By': 'Express', 'content-type': 'application/json', 'date': 'Thu, 16 Nov 2023 15:46:04 GMT', 'content-length': '799', 'connection': 'close'})
HTTP response body: {"error":"Failed to delete pipeline (v1beta1) c26ba6b0-f259-469a-836a-178e98d1de2c. Check error stack: Invalid input error: Failed to delete pipeline with id c26ba6b0-f259-469a-836a-178e98d1de2c as it has existing pipeline versions (e.g. 869951bf-9299-47a2-8d35-4d2038bcb5f6)","code":3,"message":"Failed to delete pipeline (v1beta1) c26ba6b0-f259-469a-836a-178e98d1de2c. Check error stack: Invalid input error: Failed to delete pipeline with id c26ba6b0-f259-469a-836a-178e98d1de2c as it has existing pipeline versions (e.g. 869951bf-9299-47a2-8d35-4d2038bcb5f6)","details":[{"@type":"type.googleapis.com/google.rpc.Status","code":3,"message":"Failed to delete pipeline with id c26ba6b0-f259-469a-836a-178e98d1de2c as it has existing pipeline versions (e.g. 869951bf-9299-47a2-8d35-4d2038bcb5f6)"}]}
```

These changes are noted [here](https://github.com/canonical/kfp-operators/pull/383/files#diff-cf2552fea155df80690c03878d4970941f12f251449d93388c1121ca4fcd34e2R45), for example.

### Testing instructions

#### Testing environment
* microk8s 1.25-strict/stable
  * ensure the following addons are enabled `dns hostpath-storage rbac metallb:10.64.140.43-10.64.140.49`
* juju 3.1/stable, a juju controller already bootstrapped is also required
* juju model `kubeflow`
* tox installed

Testing v1 pipelines
Execute the functional tests with: `tox -e bundle-integration-v1 -- --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2`
The above will build and deploy all kfp-operators charms from [dnplas-add-kfp-v2-integration-tests](https://github.com/canonical/kfp-operators/tree/dnplas-add-kfp-v2-integration-tests) and then execute the following test cases:
* upload v1, v2 pipeline
* create a run with a v1, v2 pipeline
* create a recurring run with a v2 pipeline
* apply a viewer
* check the viz server
 
Testing v2 pipelines
Same steps as above, but the command is `tox -e bundle-integration-v2 -- --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2`

> NOTE: bundle functional tests are expected to pass in this PR.

TODO:

- [x] provide testing instructions using tox
